File: 2015/python/17.py
# Advent of Code 2015, day 17
# (c) aichingert

import itertools

F = 150 

def solve(con,part):
    ans = 0
    for i in range(len(con)):
        if part and ans != 0:
            return ans

        for comb in itertools.combinations(con, i):
            if F == sum(comb):
                ans += 1
    return ans

con = [int(c) for c in open("../input/17").read().strip().split('\n')]

print("Part 1:",solve(con, False))
print("Part 2:",solve(con, True))

File: 2015/rust/17.rs
// Advent of Code 2015, day 17
// (c) aichingert

#[path = "../../utils/rust/combinations.rs"]
mod combinations;
use combinations::combinations;

fn solve(cont: &Vec<i32>, part: bool) -> u32 {
    let mut ans = 0;

    for s in 1..cont.len() {
        if part && ans != 0 {
            return ans;
        }

        for comb in combinations(s, cont) {
            if comb.iter().sum::<i32>() == 150 {
                ans += 1;
            }
        }
    }

    ans
}

fn main() {
    let inp = std::fs::read_to_string("../input/17").unwrap();
    let inp = inp
        .lines()
        .map(|n| n.trim().parse::<i32>().unwrap())
        .collect::<Vec<i32>>();

    println!("Part 1: {}", solve(&inp, false));
    println!("Part 2: {}", solve(&inp, true));
}
